nbs/01a_losses.ipynb

```diff
@@ -414,6 +414,113 @@
     "    def decodes(self, out):    return out.argmax(dim=-1)"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We present a general `Dice` loss for segmentation tasks. It is commonly used together with `CrossEntropyLoss` or `FocalLoss` in kaggle competations."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#export\n",
+    "def _one_hot(x, classes, axis=1):\n",
+    "    \"Target mask to one hot\"\n",
+    "    return torch.stack([torch.where(x==c, 1,0) for c in range(classes)], axis=axis)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#export\n",
+    "class DiceLoss:\n",
+    "    \"Dice loss for segmentation\"\n",
+    "    def __init__(self, axis=1, smooth=1): \n",
+    "        store_attr()\n",
+    "    def __call__(self, pred, targ):\n",
+    "        targ = _one_hot(targ, pred.shape[self.axis])\n",
+    "        pred, targ = flatten_check(self.activation(pred), targ)\n",
+    "        inter = (pred*targ).sum()\n",
+    "        union = (pred+targ).sum()\n",
+    "        return 1 - (2. * inter + self.smooth)/(union + self.smooth)\n",
+    "    \n",
+    "    def activation(self, x): return F.softmax(x, dim=self.axis)\n",
+    "    def decodes(self, x):    return x.argmax(dim=self.axis)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "dl = DiceLoss()\n",
+    "output = torch.randn(32, 4, 5, 10)\n",
+    "target = torch.randint(0,2,(32, 5, 10))\n",
+    "_ = dl(output, target)"
```

Comment:
Could you add some simple example tests that confirm the loss is correct? Currently your tests aren't actually testing anything.

Reply:
Would you mind helping me with this? I added some test for the `one_hot` and that the decodes work, but how would you unit test this? failing mostly happen with `flatten_check` so I would just be testing that.

Reply:
I'd test stuff like two masks that were identical, two which were entirely non-overlapping, and two which overlapped 50% exactly.

Reply:
good idea

Reply:
done.